### DISCUSSION TITLE
Null propagating expression of struct type requires to be wrapped in parenthesis

### DISCUSSION THREAD

**Discussion by MrTA2** [dotnet/csharplang#3956] [General] [answered]
Consider the following:

```
class A { public int X; }
A? a = GetA();
int? x = a?.X; // fine
string? s = a?.X?.ToString(); // broken
string? s2 = (a?.X)?.ToString(); // fine
```

Why is it so? The `a?.X?.ToString()` looks like valid syntax, but the compiler doesn't like it.

To make things worse, VS suggests removing parentheses from `(a?.X)?.ToString()` which is impossible, as it leads to a compilation error.

**Comment by HaloFour** [comment]
You want:

```cs
string? s = a?.X.ToString();
```

This behavior is by design.  The null-propagating operator bails immediately on the first `null`, you don't have to chain the operator to the right in the expression.  It's not possible for the intermediate value in `a?.X` to be `null` unless it's wrapped in paranthesis.

**Comment by HaloFour** [comment, marked as answer]
You _might_ need to chain if `X` was itself nullable, e.g.:

```cs
class A { public int? X }
string? s = a?.X?.ToString();
```

That basically expands into:

```cs
string? s;
if (a != null)
{
    int? temp = a.X;
    if (temp != null)
    {
        s = temp.Value.ToString();
    }
    else
    {
        s = null;
    }
}
else
{
    s = null;
}
```
